<?xml version='1.0'?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" name="worm">
 <xacro:property name="body_length" value="3" />
 <xacro:property name="axel_offset" value="0.5" />
 <xacro:property name="axel_z_offset" value="-0.15" />


 <xacro:property name="body_x" value="2" />
 <xacro:property name="body_y" value="0.5" />
 <xacro:property name="body_z" value="0.25" />
 <xacro:property name="body_mass" value="40" />

 <xacro:property name="axle_x" value="0.2" />
 <xacro:property name="axle_y" value="1" />
 <xacro:property name="axle_z" value="0.2" />
 <xacro:property name="axle_mass" value="15" />




 <xacro:macro name="wheel_bumper_macro" params="reference">
 <gazebo reference="${reference}">
   <!-- contact sensor -->
   <sensor type="contact" name="${reference}_contact_sensor">
     <update_rate>50.0</update_rate>
     <always_on>true</always_on>
     <contact>
       <collision>${reference}_collision</collision>
     </contact>
     <plugin name="lside_bumper_plugin" filename="libgazebo_ros_bumper.so">
       <bumperTopicName>bumper_${reference}</bumperTopicName>
       <frameName>World</frameName>
  </plugin>
   </sensor>
 </gazebo>
 </xacro:macro>



  <xacro:macro name="wheel_macro" params="name model">
    <link name="${name}">
     <inertial>
      <mass value="7.5"/>
      <origin rpy="0 0 0" xyz="0 0 0"/>
      <inertia ixx="0.49" ixy="0" ixz="0" iyy="0.37" iyz="0" izz="0.37"/>
     </inertial>
     <collision>
       <origin xyz="0 0 0" rpy="0 1.57079632679 0"/>
       <laser_retro>0</laser_retro>
       <max_contacts>10</max_contacts>
       <pose >0 0 0 0 0 0</pose>
       <geometry>
         <cylinder length="0.45" radius="0.28"/>
       </geometry>
     </collision>
     <visual name="visual">
      <origin rpy="0 0 1.570796" xyz="0 0 0"/>
      <geometry>
       <mesh filename="package://ros_robotics/meshes/${model}" scale="1 1 1"/>
      </geometry>
      <material name="silver">
       <color rgba="0.75 0.75 0.75 1"/>
      </material>
     </visual>
    </link>
    <xacro:wheel_bumper_macro reference="${name}"/>
  </xacro:macro>


 <xacro:macro name="sliding_joint_macro" params="xyz name parent child">
  <joint name="${name}" type="prismatic">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <axis xyz="1  0  0"/>
   <limit effort="1000" lower="-0.15" upper="0.2" velocity="0.1"/>   
  </joint>
  <xacro:trans_macro name="${name}"/>
 </xacro:macro>

 <xacro:macro name="fixed_pivot_joint_macro" params="xyz name parent child">
  <joint name="${name}" type="fixed">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
  </joint>
 </xacro:macro>

<xacro:macro name="body_macro" params="xyz name">
 <link name="${name}">
  <inertial>
   <origin xyz="${xyz}"/>
   <mass value="${body_mass}"/>
   <inertia ixx="${(1/12)*body_mass*(pow(body_z,2) + pow(body_y,2))}" ixy="0" ixz="0" iyy="${(1/12)*body_mass*(pow(body_z,2) + pow(body_x,2))}" iyz="0" izz="${(1/12)*body_mass*(pow(body_x,2) + pow(body_y,2))}"/>
  </inertial>
  <collision name="collision">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <box size="2 0.5 0.25" />
   </geometry>
  </collision>
  <visual name="visual">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <box size="2 0.5 0.25" />
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
  </link>
 </xacro:macro>

 <xacro:macro name="trans_macro" params="name">
  <transmission name="simple_trans">
   <type>transmission_interface/SimpleTransmission</type>
   <joint name="${name}">
     <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
   </joint>
   <actuator name="motor_${name}">
     <mechanicalReduction>1</mechanicalReduction>
   </actuator>
  </transmission>
 </xacro:macro>

<xacro:macro name="axle_macro" params="xyz name">
 <link name="${name}">
  <inertial>
   <mass value="${axle_mass}"/>
   <origin rpy="0  0  0" xyz="0  0  0"/>
   <inertia ixx="${(1/12)*axle_mass*(pow(axle_z,2) + pow(axle_y,2))}" ixy="0" ixz="0" iyy="${(1/12)*axle_mass*(pow(axle_z,2) + pow(axle_x,2))}" iyz="0" izz="${(1/12)*axle_mass*(pow(axle_x,2) + pow(axle_y,2))}"/>
  </inertial>
  <collision name="collision">
   <origin rpy="0  0  0" xyz="0    0   0"/>
   <geometry>
    <mesh filename="package://ros_robotics/meshes/Front_Axle.STL" scale="1.0 1.0 1.0"/>
   </geometry>
  </collision>
  <visual name="visual">
   <origin rpy="0  0  0" xyz="0 0 0"/>
   <geometry>
    <mesh filename="package://ros_robotics/meshes/Front_Axle.STL" scale="1 1 1"/>
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
 </link>
</xacro:macro>



<xacro:macro name="joint_wheel_macro" params="xyz name parent child">
 <joint name="${name}" type="revolute">
  <parent link="${parent}"/>
  <child link="${child}"/>
  <origin rpy="0    0    1.57079632679" xyz="${xyz}"/>
  <axis xyz="1  0  0"/>
  <limit effort="200" lower="-1e+16" upper="1e+16" velocity="-1"/>
 </joint>
 <xacro:trans_macro name="${name}"/>
</xacro:macro>

 <xacro:macro name="body_joint_macro" params="xyz axis name parent child">
  <joint name="${name}" type="revolute">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <axis xyz="${axis}"/>
   <limit effort="-1" lower="-1.57" upper="1.57" velocity="-1"/>
  </joint>
   <xacro:trans_macro name="${name}"/>
 </xacro:macro>

 <xacro:macro name="pivot_joint_macro" params="xyz name parent child">
  <joint name="${name}" type="revolute">
   <parent link="${parent}"/>
   <child link="${child}"/>
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <axis xyz="0 0 1"/>
   <limit effort="-1" lower="-0.18" upper="0.18" velocity="-1"/>
  </joint>
  <xacro:trans_macro name="${name}"/>
 </xacro:macro>

<xacro:macro name="dummy_link_macro" params="xyz name">
 <link name="${name}">
  <inertial>
   <mass value="0.001"/>
   <inertia ixx="0.001" ixy="0" ixz="0" iyy="0.001" iyz="0" izz="0.001"/>
  </inertial>
  <visual name="visual">
   <origin rpy="0 0 0" xyz="${xyz}"/>
   <geometry>
     <sphere radius="0.25"/>
   </geometry>
   <material name="silver">
    <color rgba="0.75 0.75 0.75 1"/>
   </material>
  </visual>
  </link>
 </xacro:macro>



<!-- body one -->
<xacro:body_macro name="body_1" xyz="0 0 0"/>

<xacro:pivot_joint_macro name="pivot_joint_1_2" parent="body_1" child="axle_1_1" xyz="${axel_offset} 0 ${axel_z_offset}"/>
<xacro:axle_macro name="axle_1_1" xyz="0 0 0"/>
<xacro:joint_wheel_macro xyz="0  0.57526 0" name="joint_wheel_1_1_1" parent="axle_1_1" child="wheel_1_1_1"/>
<xacro:joint_wheel_macro xyz="0  -0.57526 0" name="joint_wheel_1_1_2" parent="axle_1_1" child="wheel_1_1_2"/>
<xacro:wheel_macro name="wheel_1_1_1" model="frostyboy_wheel.STL"/>
<xacro:wheel_macro name="wheel_1_1_2" model="frostyboy_wheel.STL"/>

<xacro:sliding_joint_macro name="pivot_joint_1_1" parent="body_1" child="axle_1_2" xyz="${-axel_offset} 0 ${axel_z_offset}"/>
<xacro:axle_macro name="axle_1_2" xyz="0 0 0"/>
<xacro:joint_wheel_macro xyz="0  0.57526 0" name="joint_wheel_1_2_1" parent="axle_1_2" child="wheel_1_2_1"/>
<xacro:joint_wheel_macro xyz="0  -0.57526 0" name="joint_wheel_1_2_2" parent="axle_1_2" child="wheel_1_2_2"/>
<xacro:wheel_macro name="wheel_1_2_1" model="frostyboy_wheel.STL"/>
<xacro:wheel_macro name="wheel_1_2_2" model="frostyboy_wheel.STL"/>



<!-- https://answers.ros.org/question/258420/ball-joint-in-urdf/ -->
<!-- check_urdf <(xacro inorder model.urdf.xacro) inorder should have two dashes before it -->

<gazebo>
  <plugin name="gazebo_ros_control" filename="libgazebo_ros_control.so">
   <robotNamespace>/worm</robotNamespace>
   <robotSimType>gazebo_ros_control/DefaultRobotHWSim</robotSimType>
   <legacyModeNS>true</legacyModeNS>
  </plugin>
</gazebo>


</robot>
